### ISSUE TITLE
what is the detailed activation collected in 'attention' key in SDT

### ISSUE THREAD

**Issue by Rangs02245** [dubcyfor3/Prosperity#3]
Dear authors, thanks for your awesome work! It really helps to understand the whole architecture dataflow!

When looking through the code, I find it difficult to understand the network config SDT_encoder defined for Spike-Driven Transformer (SDT):

```
     ('lif_q', [dim * sequence_length * 3, batch_size, time_steps]), 
     ('attention', [dim, sequence_length, num_head, batch_size, time_steps]),
     ('fc_o', [dim, dim, sequence_length, batch_size, time_steps]),
     ('lif_o', [dim * sequence_length, batch_size, time_steps]),
     ('fc_1', [dim, dim * mlp_ratio, sequence_length, batch_size, time_steps]),
     ('lif_1', [dim * mlp_ratio * sequence_length, batch_size, time_steps]),
     ('fc_2', [dim * mlp_ratio, dim, sequence_length, batch_size, time_steps]),
     ('lif_2', [dim * sequence_length, batch_size, time_steps])
```

In SDT original code definition, the attention forward process is: 

```       
        kv = k.mul(v)
        if hook is not None:
            hook[self._get_name() + str(self.layer) + "_kv_before"] = kv
        if self.dvs:
            kv = self.pool(kv)
        kv = kv.sum(dim=-2, keepdim=True)
        kv = self.talking_heads_lif(kv)
        if hook is not None:
            hook[self._get_name() + str(self.layer) + "_kv"] = kv.detach()
        x = q.mul(kv)
        if self.dvs:
            x = self.pool(x)
        if hook is not None:
            hook[self._get_name() + str(self.layer) + "_x_after_qkv"] = x.detach()
        x = x.transpose(3, 4).reshape(T, B, C, H, W).contiguous()
        x = (
            self.proj_bn(self.proj_conv(x.flatten(0, 1)))
            .reshape(T, B, C, H, W)
            .contiguous()
        )
        x = x + identity
        return x, v, hook
```

link: https://github.com/BICLab/Spike-Driven-Transformer/blob/3faa82272eb499e4d56ea77707d4536e7c69a54b/module/ms_conv.py#L221-L244

1. if not misunderstood, q_lif, k_lif, v_lif are concatenated into lif_q? 
2. The detailed attention goes from kv mul, then row sum, them q.mul(kv), lastly goes through proj_conv and gets residual connection. I am wondering what exactly the 'attention' is. And I suppose fc_o means output from x = x + identity?

Hope for further guidance and explanation. Thanks for you nice work!

**Comment by dubcyfor3** [comment]
Hi,

Thanks for using our code and for digging carefully into the code!

The naming convention in `config.py` does not strictly correspond to the original SDT implementation. Instead, it follows the more common Transformer attention terminology, so the simulator can use a unified abstraction.

1. **Yes**, Q, K, and V and their LIF layers are concatenated together and represented as `fc_q` and `lif_q` in our config.

2. The `attention` stage in our code corresponds to the whole process that combines Q, K, and V. For SDT specifically, this includes `k.mul(v)`, the row-wise sum, and then `q.mul(kv)`. You can find the detailed simulation behavior of this stage in `simulator.py`.

3. `fc_o` corresponds to the `proj_conv` in the SDT code. The residual addition (`x + identity`) is treated as fused with the following LIF stage since its overhead is relatively lightweight in our modeling.

Hope this helps clarify the mapping!

**Comment by Rangs02245** [comment]
Thanks for clarification!

**Comment by Rangs02245** [comment]
Some updates. 
1. In `run_fc` in `simulator.py`, why only 'store all' can use n==0 to skip repeated activation load for computation in all output row $n \in N$?  from my understanding, for 'store row', GLB can accommodate an activation 'row' of size `tile_size_M × K`, so the same value should be reused for all n under the same m? By addding n==0 in activation 'store row', it is also symmetry compared with the code in weight load.

https://github.com/dubcyfor3/Prosperity/blob/6ee1c6f1cb419fcf942f2eda63db84ca28248f4b/simulator/simulator.py#L357-L375

2. I am confused about the GLB and Local buffer definition and usage, since all variables defined in simulator use `g_act`, `g_wgt`, does the g means global? if so, what's the role of local buffer, the psum? but why `g_psum`? I suppose the global and local buffer are implemeted in a higher abstract level in simulator? In `accelerator.py`, `sram_size['act'], sram_size['wgt'], sram_size['out']` are both global buffer.

3. Some errors are encountered when running PTB baselien on specific network, especially when running this line because the second dim may not be integer. I am wondering if authors met this problem before. How to deal with that? 

https://github.com/dubcyfor3/Prosperity/blob/6ee1c6f1cb419fcf942f2eda63db84ca28248f4b/simulator/baselines.py#L299

4. If not misunderstood, except PTB, other baselines do not compute mem_access, why and can you give some hints on computing others'.

Hope for further guidance. I'm new to this field, sorry for disturbance and many thanks for your patience.
